### PR TITLE
:hammer: minor refactor in trainer.

### PR DESCRIPTION
**Pull request by kartik4949**
## What does this PR do?

A better way to call train_epoch hook

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by carmocca** [comment]
If it is done at the end of the `except` and `else`, shouldn't it be instead in a `finally`?

**Comment by kartik4949** [comment]
> If it is done at the end of the `except` and `else`, shouldn't it be instead in a `finally`?

adding finally we can have two self.train_on_epoch_end executions..
hence cant add finally

**Review by Borda — APPROVED** [review]

**Review by ydcjeff — APPROVED** [review]

**Comment by carmocca** [comment]
> > If it is done at the end of the `except` and `else`, shouldn't it be instead in a `finally`?
> 
> adding finally we can have two self.train_on_epoch_end executions..
> hence cant add finally
> 
> 'else' is like finally but it executes 'else' block if everything runs fine inside 'try' block..
> @carmocca :)

How would there be two `on_train_end` executions?

Currently we have (simplified):

```python
try:
    do_stuff()
    on_train_end()
except KeyboardInterrupt:
    handle_interrupt()
    on_train_end()
```
and you proposed:
```python
try:
    do_stuff()
except KeyboardInterrupt:
    handle_interrupt()
    on_train_end()
else:
    on_train_end()
```
but why not use finally instead and have the call in one place?
```python
try:
    do_stuff()
except KeyboardInterrupt:
    handle_interrupt()
finally:
    on_train_end()
```

Notice how we call `on_train_end()` before every return within the try block. By using finally, we remove all those duplicated calls because it would execute even if there was an uncaught exception or a return statement in one of the other blocks.

So the refactored function would be:

```python
def train(self):
    self.run_sanity_check(self.get_model())

    # set stage for logging
    self.logger_connector.set_stage("train")

    self.checkpoint_connector.has_trained = False

    # enable train mode
    model = self.get_model()
    model.train()
    torch.set_grad_enabled(True)

    # reload data when needed
    self.train_loop.reset_train_val_dataloaders(model)

    # hook
    self.train_loop.on_train_start()

    try:
        if self.train_loop.should_skip_training():
            return

        # run all epochs
        for epoch in range(self.current_epoch, self.max_epochs):

            # hook
            self.train_loop.on_train_epoch_start(epoch)

            # run train epoch
            self.train_loop.run_training_epoch()

            if self.max_steps and self.max_steps <= self.global_step:
                return

            # update LR schedulers
            self.optimizer_connector.update_learning_rates(interval='epoch')

            # early stopping
            met_min_epochs = epoch >= self.min_epochs - 1
            met_min_steps = self.global_step >= self.min_steps if self.min_steps else True

            if self.should_stop:
                if met_min_epochs and met_min_steps:
                    return
                else:
                    log.info(
                        'Trainer was signaled to stop but required minimum epochs'
                        f' ({self.min_epochs}) or minimum steps ({self.min_steps}) has'
                        ' not been met. Training will continue...'
                    )

    except KeyboardInterrupt:
        rank_zero_warn('Detected KeyboardInterrupt, attempting graceful shutdown...')

        # user could press ctrl+c many times... only shutdown once
        if not self.interrupted:
            self.interrupted = True
            self._state = TrainerState.INTERRUPTED
            self.on_keyboard_interrupt()
    finally:
        # hook
        self.train_loop.on_train_end()
```

**Comment by Borda** [comment]
> How would there be two `on_train_end` executions?

@kartik4949 mind comment on @carmocca question?

**Comment by kartik4949** [comment]
Hi @carmocca @Borda 
sorry @carmocca didnt saw that you deleted the train end hook inside if statement,
I think if we delete train_end hook inside try with if statements then we are good with finally.
and finally makes a good refactor here 
thanks for your inputs :)

**Review by carmocca — CHANGES_REQUESTED** [review]
You are still mising the call before the try (put it inside the try) and the call inside the except

**Comment by kartik4949** [comment]
> You are still missing the call before the try (put it inside the try) and the call inside the except

done ..
my latest vi plugins are behaving weird with the formatting.

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
